### PR TITLE
block: Resolve relative QCOW2 backing paths

### PR DESCRIPTION
**Pull request by Rowen-Ye**
## Summary

QCOW2 backing file paths stored in image headers may be relative. Those paths should be interpreted relative to the image that references them, but the block backend opened them relative to the process working directory.

This passes the top-level image path into the QCOW2 parser and resolves relative backing paths against that image's parent directory. Recursive backing chains pass each backing image path forward so nested relative paths are resolved from the correct layer.

### CODE REVIEW

**Review by weltling — CHANGES_REQUESTED** [review]
Nice work, this is a needed fix! Before going deeper into the details, I think there is a simpler design that avoids all the API changes.

`read_link` is already imported in `mod.rs`. You can use ` proc/self/fd/{fd}` to read the image path from the open file descriptor directly inside `parse_qcow`, right before calling `BackingFile::new`. That resolves relative backing paths in place using the already open `RawFile`, so no signature changes are needed anywhere.

Why:

- No API changes
- No `new_with_path` / `new_with_optional_path` constructors
- Recursion works for free, as fd at each level resolves to its own path
- Symlinks resolve correctly
- All callers benefit automatically

The only caveat is that `read_link` can fail if file is deleted, but that just leaves the path unresolved, which is the current behavior anyway.

I'd touch upon tests separately as well, once this suggestion has been discussed.

Thanks

**Comment by Rowen-Ye** [comment]
> Nice work, this is a needed fix! Before going deeper into the details, I think there is a simpler design that avoids all the API changes.
> 
> `read_link` is already imported in `mod.rs`. You can use ` proc/self/fd/{fd}` to read the image path from the open file descriptor directly inside `parse_qcow`, right before calling `BackingFile::new`. That resolves relative backing paths in place using the already open `RawFile`, so no signature changes are needed anywhere.
> 
> Why:
> 
> * No API changes
> * No `new_with_path` / `new_with_optional_path` constructors
> * Recursion works for free, as fd at each level resolves to its own path
> * Symlinks resolve correctly
> * All callers benefit automatically
> 
> The only caveat is that `read_link` can fail if file is deleted, but that just leaves the path unresolved, which is the current behavior anyway.
> 
> I'd touch upon tests separately as well, once this suggestion has been discussed.
> 
> Thanks

Thanks, that design is much cleaner. I updated the patch to resolve the current image path inside parse_qcow() through /proc/self/fd/{fd}, so the QcowDisk/factory API changes are gone.

The recursive case now uses the same mechanism automatically because each backing layer is parsed from its own file descriptor. I also simplified the test to create the overlay with a relative backing path directly.

File: block/src/qcow/mod.rs [hunk 1]
```diff
@@ -425,9 +446,11 @@ pub(crate) fn parse_qcow(
     }
 
     let direct_io = file.is_direct();
+    let current_image_path = read_link(format!("/proc/self/fd/{}", file.as_raw_fd())).ok();
```
Comment (weltling) [inline block/src/qcow/mod.rs]: The resolution can be done in place right here, before calling `BackingFile::new`. I would suggest to mutate the backing path directly like this:

```rust
if let Some(ref mut config) = header.backing_file {
    if !Path::new(&config.path).is_absolute() {
        if let Ok(img) = read_link(format!("/proc/self/fd/{}", file.as_raw_fd())) {
            if let Some(parent) = img.parent() {
                config.path = parent.join(&config.path).to_string_lossy().into_owned();
            }
        }
    }
}

let backing_file = BackingFile::new(
    header.backing_file.as_ref(),
    direct_io,
    max_nesting_depth,
    sparse,
)?;
```

With this, `BackingFile::new` keeps its original signature and the `resolve_backing_file_path` helper function is not needed. `BackingFile::new` just opens `config.path` as before, except now the path is already absolute by the time it gets there.

Thanks

File: block/src/qcow/mod.rs [hunk 2]
```diff
@@ -204,13 +206,15 @@ impl BackingFile {
             ));
         }
 
+        let backing_path = resolve_backing_file_path(current_image_path, &config.path);
+        let backing_path_display = backing_path.display().to_string();
```
Comment (weltling) [inline block/src/qcow/mod.rs]: This goes away if the in-place mutation approach is adopted, as config.path is already a String.

Otherwise, `to_string_lossy()` and a name like `backing_path_str` would be a better fit - `display()` is meant for user facing rendering, not for storing in error variants.

Thanks

File: block/src/qcow_sync.rs [hunk 3]
```diff
@@ -658,6 +660,51 @@ mod unit_tests {
         test_backing_file_read_impl(true);
     }
 
+    #[test]
+    fn test_relative_backing_file_read() {
```
Comment (weltling) [inline block/src/qcow_sync.rs:707]: This only covers the happy path. I'd suggest to consider adding tests for error and edge cases:

- Backing file does not exist. Verify that the error message contains the resolved path (the temp dir prefix), not just the bare relative name
- Relative path with `..` components (e.g. `../sibling_dir/backing.raw`). Verify it resolves correctly from the image's parent directory
- Absolute backing path - verify it is used as is and not joined with anything
- `read_link` failure, e.g. image opened from an fd with no filesystem path. Verify graceful fallback to the unresolved relative path

Also, this test implicitly relies on the process cwd not being the temp directory. If it were, the test would pass for the wrong reason. Consider adding an assertion that the cwd differs from the temp directory, or explicitly changing to a different directory before opening the overlay.

Thanks

**Review by weltling — CHANGES_REQUESTED** [review]
Looks much better indeed, thanks for the quick turnaround on the idea! The fd
based resolution inside `parse_qcow` is exactly what I had in mind. A few
remaining items in the inline comments yet - please consider a simplification
to the resolution logic and strengthening the tests.

Thanks!

File: block/src/qcow_sync.rs (hunk 3 above)
Comment (Rowen-Ye) [reply block/src/qcow_sync.rs:707]: Thanks, I updated the patch following your suggestion.

The resolution now happens in-place inside `parse_qcow()` before calling
`BackingFile::new()`, so `BackingFile::new()` keeps its original signature and
the helper function is gone.

I also strengthened the tests to cover:

- relative backing paths resolved from the overlay image directory
- missing relative backing files reporting the resolved path
- relative paths with `..` components
- absolute backing paths being used as-is
- fallback behavior for an fd without a filesystem path

Thanks again for the guidance.

File: block/src/qcow_sync.rs [hunk 4]
```diff
@@ -658,6 +666,184 @@ mod unit_tests {
...
+
+        let buf = async_read(&disk, 0, cluster_size as usize);
+        assert_eq!(
+            &buf[..],
+            &pattern[..cluster_size as usize],
+            "Absolute backing file path should be used as is"
+        );
+    }
+
+    #[cfg(target_os = "linux")]
```
Comment (weltling) [inline block/src/qcow_sync.rs]: nit: This is not needed, Cloud Hypervisor is Linux only.

Thanks

**Review by weltling — APPROVED** [review]
Looks good! The resolution logic is clean, tests are comprehensive, and `BackingFile::new` keeps its original signature. Nice use of let chains for the
flat condition.

Thanks for iterating on this!

File: block/src/qcow_sync.rs [hunk 5]
```diff
@@ -658,6 +666,184 @@ mod unit_tests {
...
+            .truncate(true)
+            .open(overlay_path)
+            .unwrap();
+        let header =
+            QcowHeader::create_for_size_and_path(3, file_size, Some(backing_path)).unwrap();
+        header.write_to(&mut file).unwrap();
+        file.sync_all().unwrap();
+    }
+
+    #[cfg(target_os = "linux")]
```
Comment (phip1611) [inline block/src/qcow_sync.rs]: same here - not needed

Comment (Rowen-Ye) [reply block/src/qcow_sync.rs]: Thanks, agreed.  I have removed those cfg guards.

File: block/src/qcow_sync.rs (hunk 4 above)
Comment (Rowen-Ye) [reply block/src/qcow_sync.rs]: Thanks, agreed.  I have removed those cfg guards.

File: block/src/qcow/mod.rs [hunk 6]
```diff
@@ -245,7 +247,7 @@ impl BackingFile {
                         let source = e
                             .into_source()
                             .and_then(|s| s.downcast::<Error>().ok())
-                            .map(|qcow_err| Error::BackingFileOpen(config.path.clone(), qcow_err));
+                            .map(|qcow_err| Error::BackingFileOpen(backing_path.clone(), qcow_err));
```
Comment (rbradford) [inline block/src/qcow/mod.rs]: This refactor only change seems independent of the other changes - should be a in a separate commit.

File: block/src/qcow_sync.rs [hunk 7]
```diff
@@ -658,6 +664,182 @@ mod unit_tests {
...
     }
 
+    fn create_raw_backing(path: &Path, pattern: &[u8]) {
+        let mut backing_file = File::create(path).unwrap();
+        backing_file.write_all(pattern).unwrap();
+        backing_file.sync_all().unwrap();
+    }
+
+    fn create_qcow2_overlay(overlay_path: &Path, backing_path: &str, file_size: u64) {
+        {
```
Comment (rbradford) [inline block/src/qcow_sync.rs]: What's with the extra scope?

File: block/src/qcow/mod.rs [hunk 8]
```diff
@@ -425,9 +427,20 @@ pub(crate) fn parse_qcow(
     }
 
     let direct_io = file.is_direct();
+    let backing_file_config = header.backing_file.as_ref().map(|config| {
```
Comment (rbradford) [inline block/src/qcow/mod.rs:432]: This definitely should be commented or maybe even a function (with a unit test.)

File: block/src/qcow_sync.rs [hunk 9]
```diff
@@ -658,6 +664,182 @@ mod unit_tests {
...
+
+    fn create_memfd(name: &str) -> File {
+        let name = std::ffi::CString::new(name).unwrap();
+        // SAFETY: memfd_create is a self-contained syscall; on success it
+        // returns a new fd owned by this function.
+        let fd = unsafe { libc::syscall(libc::SYS_memfd_create, name.as_ptr(), 0) };
+        assert!(fd >= 0, "memfd_create failed");
+        // SAFETY: the successful memfd_create call returned a valid owned fd.
+        unsafe { File::from_raw_fd(fd as RawFd) }
+    }
```
Comment (rbradford) [inline block/src/qcow_sync.rs]: Must already have this somewhere in block/ no?

**Review by rbradford — CHANGES_REQUESTED** [review]

File: block/src/qcow_sync.rs (hunk 9 above)
Comment (Rowen-Ye) [reply block/src/qcow_sync.rs]: Thanks for the review. I addressed these by removing the unrelated BackingFile::new() refactor, adding a comment explaining the fd-based backing path resolution, dropping the unnecessary scope in the test helper, and replacing the local memfd helper with TempFile::into_file() for the no-filesystem-path fallback case.

File: block/src/qcow/mod.rs [hunk 10]
```diff
@@ -425,9 +426,23 @@ pub(crate) fn parse_qcow(
     }
 
     let direct_io = file.is_direct();
+    // QCOW2 relative backing paths are resolved from the image that stores
+    // them. Resolve only the config passed to BackingFile::new(), leaving the
+    // header copy unchanged so the original backing filename is preserved.
+    let backing_file_config = header.backing_file.as_ref().map(|config| {
+        let mut config = config.clone();
+        if !Path::new(&config.path).is_absolute()
+            && let Ok(img) = read_link(format!("/proc/self/fd/{}", file.as_raw_fd()))
```
Comment (rbradford) [inline block/src/qcow/mod.rs]: Nice use of let chain! I was thinking maybe we could have a better variable name than `img` but i'm struggling for inspiration.

**Review by rbradford — APPROVED** [review]

File: block/src/qcow/mod.rs (hunk 10 above)
Comment (rbradford) [reply block/src/qcow/mod.rs]: Can we s/img/disk_path/

Comment (Rowen-Ye) [reply block/src/qcow/mod.rs]: Thanks, good suggestion. I renamed it to disk_path.

**Review by rbradford — APPROVED** [review]
